os/binfmt/binfmt_loadlibrary.c

```diff
@@ -0,0 +1,148 @@
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <string.h>
+#include <debug.h>
+#include <errno.h>
+
+#include <tinyara/kmalloc.h>
+#include <tinyara/sched.h>
+#include <tinyara/binfmt/binfmt.h>
+#include <tinyara/binary_manager.h>
+
+#include "binfmt.h"
+#include "binary_manager/binary_manager.h"
+#include <tinyara/mm/mm.h>
+
+#if defined(CONFIG_BINFMT_ENABLE) && defined(CONFIG_SUPPORT_COMMON_BINARY)
+
+struct binary_s *g_lib_binp;
+uint32_t *g_umm_app_id;
+
+#ifdef CONFIG_ARMV7M_MPU
+extern uint32_t g_mpu_region_nr;
+void mpu_configure_library_region(uint32_t region, uintptr_t base, size_t size, uint8_t readonly, uint8_t execute);
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: load_library
+ *
+ * Description:
+ *   This is a convenience function that wraps load_module and other steps into
+ *   one call and is used to load a library and export its symbols.
+ *
+ * Input Parameters:
+ *   filename  - The path to the program to be executed. If
+ *               CONFIG_LIB_ENVPATH is defined in the configuration, then
+ *               this may be a relative path from the current working
+ *               directory. Otherwise, path must be the absolute path to the
+ *               program.
+ *
+ * Returned Value:
+ *   This is an end-user function, so it follows the normal convention:
+ *   It returns the PID of the exec'ed module.  On failure, it returns
+ *   -1 (ERROR) and sets errno appropriately.
+ *
+ ****************************************************************************/
+int load_library(FAR const char *filename)
+{
+	FAR struct binary_s *bin = NULL;
+	int errcode;
+	int ret;
+
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	bin = g_lib_binp;
+	/* If we find a non-null value for bin, it means that
+	 * we are in a reload scenario.
+	 */
+	if (bin) {
+		if (!bin->data_backup) {
+			errcode = -EINVAL;
```

Comment:
Sometimes, you set the errno with negative value, but sometimes with positive.
Please check this.

Reply:
fixed.